src/goto-programs/goto_inline.cpp

```diff
@@ -375,231 +167,90 @@ Function: replace_location
-    parameter_assignments(target->source_location, identifier, f.type, arguments, tmp);
-    tmp.destructive_append(tmp2);
-    parameter_destruction(target->source_location, identifier, f.type, tmp);
+      const goto_programt &goto_program=goto_function.body;
 
-    if(f.is_hidden())
-    {
-      source_locationt new_source_location=
-        function.find_source_location();
-
-      if(new_source_location.is_not_nil())
+      if(goto_function.is_inlined() ||
+         goto_program.instructions.size()<=smallfunc_limit)
       {
-        new_source_location.set_hide();
-
-        Forall_goto_program_instructions(it, tmp)
-        {
-          if(it->function==identifier)
-          {
-            // don't hide assignment to lhs
-            if(it->is_assign() && to_code_assign(it->code).lhs()==lhs)
-            {
-            }
-            else
-            {
-              replace_location(it->source_location, new_source_location);
-              replace_location(it->guard, new_source_location);
-              replace_location(it->code, new_source_location);
-            }
-
-            it->function=target->function;
-          }
-        }
+        assert(goto_inlinet::is_call(i_it));
+        call_list.push_back(goto_inlinet::callt(i_it, false));
       }
     }
-
-    // set up location instruction for function call
-    target->type=LOCATION;
-    target->code.clear();
-
-    goto_programt::targett next_target(target);
-    next_target++;
-
-    dest.instructions.splice(next_target, tmp.instructions);
-    target=next_target;
-
-    recursion_set.erase(identifier);
   }
-  else // no body available
+  
+  try
   {
-    if(no_body_set.insert(identifier).second)
-    {
-      warning().source_location=function.find_source_location();
-      warning() << "no body for function `" << identifier << "'" << eom;
-    }
-
-    goto_programt tmp;
-
-    // evaluate function arguments -- they might have
-    // pointer dereferencing or the like
-    forall_expr(it, arguments)
-    {
-      goto_programt::targett t=tmp.add_instruction();
-      t->make_other(code_expressiont(*it));
-      t->source_location=target->source_location;
-      t->function=target->function;
-    }
-
-    // return value
-    if(lhs.is_not_nil())
-    {
-      side_effect_expr_nondett rhs(lhs.type());
-      rhs.add_source_location()=target->source_location;
-
-      code_assignt code(lhs, rhs);
-      code.add_source_location()=target->source_location;
-
-      goto_programt::targett t=tmp.add_instruction(ASSIGN);
-      t->source_location=target->source_location;
-      t->function=target->function;
-      t->code.swap(code);
-    }
-
-    // now just kill call
-    target->type=LOCATION;
-    target->code.clear();
-    target++;
-
-    // insert tmp
-    dest.instructions.splice(target, tmp.instructions);
+    goto_inline.goto_inline(inline_map, false);
+  }
+  catch(const char *e)
+  {
+    goto_inline.error() << e << messaget::eom;
+    throw 0;
```

Comment:
Applies above as well: calling error() here makes it impossible to include/report a source location. Couldn't you invoke error() at place where an exception is thrown at present, and then `throw 0;` in those places?